drivers/include/mtd.h

```diff
@@ -44,11 +44,25 @@
  *
  *   Pages are a subdivision of sectors.
  *
+ * * The **write size** is the minimum size of writes to the device, and also
+ *   the required alignment of writes.
+ *
+ *   The write size is a divider of the page. It is often between 1 to 4 bytes
+ *   long, but may be up to the full page size.
+ *
  * * The device's **flags** indicate features, eg. whether a memory location
  *   can be overwritten without erasing it first.
  *
- * Note that some properties of the backend are currently not advertised to the
- * user (see the documentation of @ref mtd_write).
+ * Unless a flag (such as @ref MTD_DRIVER_FLAG_DIRECT_WRITE or @ref
+ * MTD_DRIVER_FLAG_CLEARING_OVERWRITE) allows it, this MTD API does not allow
+ * memory areas to be written to twice between erase operations. Drivers are
+ * not expected to count write accesses, and neither do this module's
+ * functions: The performance impact would be too great. It is up to the
+ * application to only write to erased memory once. Failure to do so may damage
+ * hardware.
```

Comment:
Actually, for the MTD backend of vcdiff I'm planning to write to one page twice without erase cycles.

Example with `write_size = 4`:
1) Erase the flash
2) Write the first 2 Bytes of the first page:
```c
uint8_t buf[4];
mtd_read_page(mtd, buf, 0, 0, 4);
buf[0] = 0;
buf[1] = 1;
mtd_write_page_raw(mtd, buf, 0, 0, 4);
```
3) Write the third Byte of the first page:
```c
uint8_t buf[4];
mtd_read_page(mtd, buf, 0, 0, 4);
buf[2] = 2;
mtd_write_page_raw(mtd, buf, 0, 0, 4);
```

This should result into `00 01 02 ff`.

Is this invalid usage of MTD? This paragraph implies that, if I get it right.

Reply:
Another use case: The RIOTBOOT magic (first 4 bytes) ist written after the whole image is in place and verified. This requires the first 4 bytes written as the last step during firmware upgrade.

Reply:
That is valid use on MTD devices that advertise that zeroing overwrites can be done using the MTD_DRIVER_FLAG_CLEARING_OVERWRITE bit. (It is also valid use on MTDs that set the MTD_DRIVER_FLAG_DIRECT_WRITE flag, but then the user must really read before write and can't just replace your step 3 with "write ff ff 02 ff" and expect the same outcome, which they may with MTD_DRIVER_FLAG_CLEARING_OVERWRITE).

Initially, that flag is not set on by MTD implementation; if you know of hardware that can do it, we could add it right away. Note that the flashpage MTD backend may or may not allow this use, but is not modeled to the same granularity (i.e., it has no corresponding flag that the mtd_flashpage initialization could set). For example, I know that the older ESP32 allow this use (and thus should have this flag set), whereas nRF52 and the larger STM32 (esp. L5) do not.

Reply:
But wouldn't that mean SUIT updates can't work on nRF52 and STM32L5?

Reply:
> Another use case: The RIOTBOOT magic (first 4 bytes) ist written after the whole image is in place and verified. This requires the first 4 bytes written as the last step during firmware upgrade.

But in that case its not actually RIOTBOOT magic, but `write_size` that is written, the first FLASHPAGE_WRITE_BLOCK_ALIGNMENT are skipped while the content of `FLASHPAGE_WRITE_BLOCK_ALIGNMENT` - `write_size` are saved in a buffer so that the whole `FLASHPAGE_WRITE_BLOCK_ALIGNMENT,` is then written, so there is no overwrite happening.

Reply:
> Actually, for the MTD backend of vcdiff I'm planning to write to one page twice without erase cycles.

I think for `vcdiff` you would need to buffer writes up to `write_size`, same thing that is done for `flashwrite`

Reply:
(Oh, I missed the second comment by jue89, and simultaneously posted with fjmolinas...)

If SUIT already works in write block alignments, then great. (If it were to switch to MTD it couldn't rely on these things being build-time any more, so going by flashpage is practical). Just beware that the alignment might at one point exceed the riotboot header, leaving part of the actual firmware to be written late-but-after-the-checksome-is-written. (But likely there is no harm, for these large alignments come from ECC that would rather fail than give incomplete data).

Reply:
[meta] I consider this thread purely informational, as it is about implications not of the PR itself but of the underlying observations and their effect on flashpage users.

In terms of comments, there is the open discussion with Ben on whether RMW implementations would set write size to 1 or not, but if that is resolved and someone ACKs the PR, I'd go ahead with it even if this thread has not come to any conclusion, as long as nobody says otherwise.

Reply:
> [meta] I consider this thread purely informational, as it is about implications not of the PR itself but of the underlying observations and their effect on flashpage users.

Sure. I complete this discussion with the mandatory catch phrase: *MTD is confusing* :rofl: